src/feat/signal-distort.cc

```diff
@@ -72,5 +72,97 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+  KALDI_ASSERT(noise_eg.NumCols() == input_cols);
+
+  // According to the rows of noise_eg, form the noise_mat
+  // repeat the noise_eg blocks to have a new block which is longer than input_eg
+ 
+  // As the noise_eg is very huge and the input_eg is small normally,
+  // so we'd better not reload the "noise_eg" matrix
+  // select the noise range
+
+  Matrix<BaseFloat> selected_noise_mat;
+  selected_noise_mat.Resize(input_rows, input_cols);
+  
+  int32 noise_rows = noise_eg.NumRows();
+  int32 start_row_ind = RandInt(0, noise_rows - input_rows);
+  
+  if (noise_eg.NumRows() < input_rows) {
+    int32 indices[input_rows];
+    for (int32 i=0; i < input_rows; ++i) {
+      indices[i] = (start_row_ind + i) % noise_eg.NumRows();
+    }
+    selected_noise_mat.CopyRows(noise_eg, indices);
+  } else {
+    selected_noise_mat.AddMat(1.0, noise_eg.Range(start_row_ind, input_rows,
+                                                  0, input_cols));
+  }
+
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 0.0);
+  double input_energy = input_energy_mat.Sum();
+  Matrix<BaseFloat> noise_energy_mat(input_rows, input_cols);
+  noise_energy_mat.AddMatMatElements(1.0, selected_noise_mat, selected_noise_mat, 0.0);
+  double noise_energy = noise_energy_mat.Sum();
+
+  // In Energy domain, SNR=20log10(S/N). 
+  // 10^(SNR/20) = input_energy / (scale^2 * noise_energy)
+  double scale = input_energy / noise_energy / (pow(10,opts_.snr/20));
+  scale = sqrt(scale);
+  
+  // Add noise mat to input_eg mat
+  perturbed_eg->Resize(input_rows, input_cols);
+  perturbed_eg->CopyFromMat(input_eg);
+  perturbed_eg->AddMat(scale, selected_noise_mat);
+}
+
+void PerturbXvectorSignal::ApplyDistortion(const MatrixBase<BaseFloat> &input_egs,
+                                           Matrix<BaseFloat> *perturb_egs) {
+    // conduct ApplyAdditiveNoise
+  if (!opts_.add_noise.empty()) {
+    // choose a noise from the noise.scp/ark
+    // 1) we need to record the keys of noise_egs
+    std::vector<std::string> list_noise_egs;
+    SequentialBaseFloatMatrixReader noise_seq_reader(opts_.add_noise);
+    for (; !noise_seq_reader.Done(); noise_seq_reader.Next()) {
+      std::string key = noise_seq_reader.Key();
+      list_noise_egs.push_back(key);
+    }
+    noise_seq_reader.Close();
+
+    // 2) we random choose an noise example
+    int32 num_noise_egs = list_noise_egs.size();
+    int32 index_noise_eg = RandInt(0, num_noise_egs - 1);
+    std::string key_noise_eg = list_noise_egs[index_noise_eg];
+    RandomAccessBaseFloatMatrixReader noise_random_reader(opts_.add_noise);
+    Matrix<BaseFloat> noise_eg_mat = noise_random_reader.Value(key_noise_eg);
+    SetNoiseEgs(noise_eg_mat);
+
+    ApplyAdditiveNoise(input_egs, *noise_egs_, perturb_egs);
+    // conduct others
+    // TODO
+  } else { // deal with the opts_.noise_egs situation
```

Comment:
no need for else condition!
We can compose several different perturbation to generated perturbed_egs.